engine/audiomixer.cpp

```diff
@@ -0,0 +1,167 @@
+#include "audiomixer.h"
+
+#include "config.h"
+#include "logger.h"
+#include "pong.h"
+#include "sound.h"
+
+#include <glm/glm.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <chrono>
+#include <iostream>
+
+namespace pong
+{
+
+static constexpr int SAMPLE_RATE = 44100;
+static constexpr int FRAMES_PER_BUFFER = 256;
+static constexpr glm::vec3 LEFT_LISTENER_POSITION = glm::vec3(-1125.0f, 0.0f, 0.0f);
+static constexpr glm::vec3 RIGHT_LISTENER_POSITION = glm::vec3(1125.0f, 0.0f, 0.0f);
+static constexpr float MIN_DISTANCE = 250.0f;
+
+int AudioCallbackWrapper(const void* inputBuffer, void* outputBuffer,
+                  unsigned long framesPerBuffer,
+                  const PaStreamCallbackTimeInfo* timeInfo,
+                  PaStreamCallbackFlags statusFlags,
+                  void* userData)
+{
+    return Pong::GetInstance().GetAudioMixer().AudioCallback(inputBuffer, outputBuffer, framesPerBuffer, timeInfo, statusFlags, userData);
+}
+
+bool AudioMixer::Init()
+{
+    mVolume = Config::GetValue("volume", 0.5f);
+    LogInfo("Volume: {}", mVolume);
+
+    mSpatialAudioEnabled = Config::GetValue("spatial_audio", false);
+    LogInfo("Spatial Audio: {}", mSpatialAudioEnabled ? "Enabled" : "Disabled");
+
+    PaError err = Pa_Initialize();
+    if (err != paNoError) {
+        LogError("PortAudio initialization failed: {}", Pa_GetErrorText(err));
+    }
+
+    PaStreamParameters outputParameters;
+    outputParameters.device = Pa_GetDefaultOutputDevice();
+    outputParameters.channelCount = 2; // Stereo
+    outputParameters.sampleFormat = paFloat32;
+    outputParameters.suggestedLatency = Pa_GetDeviceInfo(outputParameters.device)->defaultLowOutputLatency;
+    outputParameters.hostApiSpecificStreamInfo = nullptr;
+
+    err = Pa_OpenStream(&mStream, nullptr, &outputParameters, SAMPLE_RATE, FRAMES_PER_BUFFER, paClipOff, AudioCallbackWrapper, nullptr);
+    if (err != paNoError) {
+        LogError("PortAudio stream opening failed: {}", Pa_GetErrorText(err));
+        Pa_Terminate();
+        return false;
+    }
+
+    err = Pa_StartStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream starting failed: {}", Pa_GetErrorText(err));
+        Pa_CloseStream(mStream);
+        Pa_Terminate();
+        return false;
+    }
+
+    return true;
+}
+
+void AudioMixer::Cleanup()
+{
+    PaError err = Pa_StopStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream stopping failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    err = Pa_CloseStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream closing failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    Pa_Terminate();
+}
+
+void AudioMixer::PlaySound(const Sound& sound)
+{
+    mPlayingSounds.emplace_back(sound);
+}
+
+void AudioMixer::PlaySound(const Sound& sound, const glm::vec3& position)
+{
+    mPlayingSounds.emplace_back(sound, position);
```

Comment:
I'll fix this when next I change things for this PR. I don't think this is quite thread safe right now. I tried to look into using a thread sanitizer, but it seems msvc has no support for one at all 😞

Reply:
embarrassing for msvc. Very unfortunate.

You may be able to compile with llvm on windows?? Not sure...

If mPlayingSounds is being used here and in the audio thread, it is true that this is not thread safe. The naive approach that you could start with is just adding a lock, but as I've long ranted about that isn't realtime safe ;)

Reply:
Ok, talked to dave today and brought this up. There is a likely simple way to get these to be realtime safe.

Basically, you need a realtime safe queue (which you already have) but between your main thread and your audio thread. This queue would be full of sound event objects. Things like "StartPaddleSound", "StopPaddleSound" etc

In the realm of your audio thread, you have a static array of all the sounds you wish to play, and on a StartPaddleSound, you simply play the paddle sound file. 

The only thing that crosses the thread boundaries is the event, and that is thread safe in your lock free queue. The sound thread deals with interpreting the event, then choosing and playing the right sound until it hears the "stop" event.


You do not have to create this now, just writing this here for posterity if you ever feel a need to circle back to implement this. This is more akin to what normal game engines do, I would think :)